autoprecompiles/src/memory_optimizer.rs

```diff
@@ -4,62 +4,59 @@ use std::hash::Hash;
 
 use itertools::Itertools;
 use powdr_constraint_solver::boolean_extractor::{self, RangeConstraintsForBooleans};
-use powdr_constraint_solver::constraint_system::{ConstraintRef, ConstraintSystem};
+use powdr_constraint_solver::constraint_system::{BusInteraction, ConstraintRef, ConstraintSystem};
 use powdr_constraint_solver::indexed_constraint_system::IndexedConstraintSystem;
-use powdr_constraint_solver::quadratic_symbolic_expression::RangeConstraintProvider;
 use powdr_constraint_solver::quadratic_symbolic_expression::{
-    NoRangeConstraints, QuadraticSymbolicExpression,
+    QuadraticSymbolicExpression, RangeConstraintProvider,
 };
 use powdr_constraint_solver::utils::possible_concrete_values;
-use powdr_number::{FieldElement, LargeInt};
-
-use crate::legacy_expression::{AlgebraicExpression, AlgebraicReference};
-use crate::optimizer::algebraic_to_quadratic_symbolic_expression;
-use crate::{SymbolicBusInteraction, SymbolicConstraint, SymbolicMachine};
+use powdr_number::FieldElement;
 
 /// The memory address space for register memory operations.
 const REGISTER_ADDRESS_SPACE: u32 = 1;
 
 /// Optimizes bus sends that correspond to general-purpose memory read and write operations.
 /// It works best if all read-write-operation addresses are fixed offsets relative to some
 /// symbolic base address. If stack and heap access operations are mixed, this is usually violated.
-pub fn optimize_memory<T: FieldElement>(
-    mut machine: SymbolicMachine<T>,
+pub fn optimize_memory<T: FieldElement, V: Hash + Eq + Clone + Ord + Display>(
+    mut system: ConstraintSystem<T, V>,
     memory_bus_id: u64,
-) -> SymbolicMachine<T> {
+    range_constraints: impl RangeConstraintProvider<T, V> + Clone,
```

Comment:
In practice, we always pass `NoRangeConstraints`. Can we remove this argument here and in the functions it's passed to then? In which scenario would we have range constraints?

Reply:
I think we should always have range constraints. I mean we can derive them via the bus interactions, why shouldn't we supply them?